lib/streamlit/caching/cache_utils.py

```diff
@@ -228,6 +329,106 @@ def _show_cached_st_function_warning(
             dg.exception(e)
 
 
+"""
+Note [DeltaGenerator method invocation]
+There are 3 different ways an st function can be invoked:
+1. Implicitly on the main DG instance (plain `st.foo` calls)
+2. Implicitly on an active context's block (`st.foo` within a `with st.block` context)
+3. Explicitly on a DG instance (`st.sidebar.foo`, `my_container.foo`)
+
+To simplify replaying messages from a cached function result, we convert all of these
+to explicit invocations. How they get rewritten depends on both implicit vs explicit,
+and if the target DG is already known within the replay.
+
+Implicit invocation on a known DG -> Explicit invocation on that DG
+Implicit invocation on an unknown DG -> Rewrite as explicit invocation on main
+    This is situation 2 above, and the DG is a block entirely outside our function call,
+    so we interpret it as "put these elements within the current context", which is achieved
```

Comment:
When you call st.foo, it is technically a call on the main DG. But if you're in a `with st.some_block_thing` block, main will look up the DG for that and use it instead of itself for handling the call. So `st.foo` actually means "call this on the dg of the current block context", (treating main as an all-encompassing block context, because it kinda is)